File: solutions/cpp/say/1/say.cpp
#include "say.h"
#include <array>
#include <cmath>
#include <stdexcept>
#include <string>

namespace say {

const std::string ones_and_teens[] = {
    "zero",
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine",
    "ten",
    "eleven",
    "twelve",
    "thirteen",
    "fourteen",
    "fifteen",
    "sixteen",
    "seventeen",
    "eighteen",
    "nineteen"
};

const std::string tens[] = {
    "",
    "ten",
    "twenty",
    "thirty",
    "forty",
    "fifty",
    "sixty",
    "seventy",
    "eighty",
    "ninety"
};

const std::string powers[] = {
    "",
    "thousand",
    "million",
    "billion"
};

std::string tail(unsigned long int remainder, const std::string& separator) {
    if (remainder)
        return separator + in_english(remainder);
    return "";
}

std::string large_n_in_english(long int n, long int divisor, const std::string& qualifier) {
    auto quantity = n / divisor;
    auto remainder = n % divisor;
    return in_english(quantity) + " " + qualifier + tail(remainder, " ");
}

std::string very_large_n_in_english(long int n) {
    unsigned int magnitude = log10(n)/3;
    if (magnitude < sizeof(powers)/sizeof(powers[0]))
        return large_n_in_english(n, pow(1000, magnitude), powers[magnitude]);
    throw std::domain_error("Out of range.");    
}

std::string in_english(long int n) {
    if (n < 0)
        throw std::domain_error("Must not be negative.");
    if (n < 20)
        return ones_and_teens[n];
    if (n < 100) 
        return tens[n / 10] + tail(n % 10, "-");
    if (n < 1000)
        return large_n_in_english(n, 100, "hundred");
    return very_large_n_in_english(n);
}

}  // namespace say

File: solutions/clojure/say/1/src/say.clj
(ns say (:require [clojure.math :as math]))


(def ones-and-teens ["", 
            "one",
            "two",
            "three",
            "four",
            "five",
            "six",
            "seven",
            "eight",
            "nine",
            "ten",
            "eleven",
            "twelve",
            "thirteen",
            "fourteen",
            "fifteen",
            "sixteen",
            "seventeen",
            "eighteen",
            "nineteen"])

(def tens ["",
           "ten",
           "twenty",
           "thirty",
           "forty",
           "fifty",
           "sixty",
           "seventy",
           "eighty",
           "ninety"])

(def thousand-powers ["",
                "thousand",
                "million",
                "billion"])

(defn join [sep s1 s2]
  (if (> (count s2) 0) 
    (apply str [s1, sep, s2])
    s1
    )
  )

(declare say)

(defn say-with-scale [num scale scale-word]
  (let [head (say (int (/ num scale))) 
        tail (say (mod num scale))]
    (if (> (count head) 0)
      (join " " (str head " " scale-word) tail) 
      tail
      )
    )
  )

(defn say-big [num]
  (cond
    (< num 1000) (say-with-scale num 100 "hundred")
    (< num (math/pow 1000 4))
      (let [order (int (/ (math/log10 num) 3))
            scale (bigint (math/pow 1000 order))
            scale-word (get thousand-powers order)] 
          (say-with-scale num scale scale-word)
        )
    :else (throw (IllegalArgumentException. "out of range"))
    )
  )

(defn say [num]
  (cond
    (< num 0) (throw (IllegalArgumentException. "out of range"))
    (< num 20) (get ones-and-teens num)
    (< num 100) (join "-" (get tens (int (/ num 10))) (say (mod num 10)))
    :else (say-big num)
    )
  )

(defn number [num] ;; <- arglist goes here
  (if (= num 0) "zero" (say num))
  )
